tflite2onnx/op/reshape.py

```diff
@@ -72,17 +72,143 @@ def parse(self):
 
         self.setParsed()
 
+    def preserveInputSpatialSemantic(self):
+        # If the input (or output) tensor of Reshape
+        # holds a special data layout semantic,
+        # we need to insert Transpose before (or after) the Reshape operator
+        # https://github.com/jackwish/tflite2onnx/issues/28
+        # An example for inserting `Transpose` before `Reshape`
+        #    ------
+        #   | Conv |
+        #    ------
+        #      |
+        #      |  to_transpose (Original input of `Reshape`)
+        #      |  (e.g. NCHW)
+        #   -------
+        #  | Trans |  e.g. perm: (0, 2, 3, 1)
+        #   -------
+        #      |
+        #      |  transposed (New created tensor)
+        #      |  (e.g. NHWC)
+        #   --------
+        #  | Reshape |
+        #   --------
+        #
+
+        assert(self.status.parsed)
+        to_transpose = self.inputs[0]
+
+        transposed_name = 'TFLITE2ONNX_Transposed_%s' % to_transpose.name
+        transposed = self.TFactory.getWithRef(to_transpose, transposed_name, True)
+
+        # Copy the layout of original input of `Reshape`
+        # e.g. current: NCHW, (source: NHWC, target: NCHW)
+        layout = copy.deepcopy(to_transpose.layout)
+        # Exchange the source and target layout for layout preservation
+        layout.current = layout.source
+        layout.source = layout.target
+        layout.target = layout.current
+        transposed.shape = layout.transform(to_transpose.shape)
+        transposed.setParsed()
+
+        # Construct the additional transpose before `Reshape`
+        trans = Transpose(self.TFactory, -1)
+        # e.g. From NCHW to NHWC
+        trans.attrs['perm'] = layout.perm
+
+        trans.inputs.append(to_transpose)
+        to_transpose.replaceConsumer(self, trans)
+        self.replaceInput(to_transpose, transposed)
+
+        trans.outputs.append(transposed)
+        transposed.addProducer(trans)
+        transposed.addConsumer(self)
+        trans.setParsed()
+
+        self.pre.append(trans)
+
+    def preserveOutputSpatialSemantic(self):
+        # If the input (or output) tensor of Reshape
+        # holds a special data layout semantic,
+        # we need to insert Transpose before (or after) the Reshape operator
+        # https://github.com/jackwish/tflite2onnx/issues/28
+
+        # An example for inserting `Transpose` after `Reshape`
+        #   -------
+        # | Reshape |
+        #   -------
+        #      |
+        #      |  to_transpose (New created tensor)
+        #      |  (e.g. NHWC)
+        #   -------
+        #  | Trans |  e.g. perm: (0, 3, 1, 2)
+        #   -------
+        #      |
+        #      |  transposed (Original `Reshape` output)
+        #      |  (e.g. NCHW)
+        #    ------
+        #   | Conv |
+        #    ------
+        #
+
+        assert(self.status.parsed)
+        transposed = self.outputs[0]
+
+        transposed_new_name = 'TFLITE2ONNX_Transposed_%s' % transposed.name
```

Comment:
This seems to be a bit hacky, maybe we can just name the new tensor as something like `'TFLITE2ONNX_ToTransposed_%s' % transposed.name`?